### ISSUE TITLE
Extend index_reduce to allow `index` to be of type List[List[int]]

### ISSUE THREAD

**Issue by KukumavMozolo** [pytorch/pytorch#130525]
### 🚀 The feature, motivation and pitch

I am looking for a memory efficient way to compute multiple averages over (variable length) subsets of large matrices.
The current implementation `index_reduce` requires `index` to be of type `Tensor` therefore one can only select uniformly fixed length subsets.
I propose to soften the constrain `index` to be of type List[List[int]] e.g. `index = [[0,1], [2,3,4], [0,5]]` so that 
`x.index_reduce_(0, index, x, 'mean')` would compute the following averages of x:
` torch.mean(x[[0,1],:],dim=0)`,` torch.mean(x[[2,3,4],:],dim=0)`,` torch.mean(x[[0,5],:],dim=0)`
